[CALCITE-2164] Fix alerts raised by lgtm.com

1m2c3t4 opened the pull request:
This PR addresses some of the alerts raised by lgtm.com (https://lgtm.com/projects/g/apache/calcite/alerts?mode=list). The alerts are of six kinds as listed in https://issues.apache.org/jira/browse/CALCITE-2164

julianhyde commented on core/src/main/java/org/apache/calcite/sql/SqlLiteral.java:
```diff
@@ -717,7 +717,7 @@ public void unparse(
       break;
     default:
-      writer.literal(value.toString());
+      writer.literal(String.valueOf(value));
```
over-cautious. we know that value is not null.

julianhyde commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java:
```diff
@@ -530,7 +530,7 @@ public static RelDataType createJoinType(
         == (systemFieldList.size()
         + leftType.getFieldCount()
-        + rightType.getFieldCount()));
+        + (rightType == null ? 0 : rightType.getFieldCount())));
```
we require that rightType is not null. add precondition, rather than making code more defensive.

julianhyde commented on druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java:
```diff
@@ -774,12 +774,13 @@ protected QuerySpec getQuery(RelDataType rowType, RexNode filter, List<RexNode>
         writeFieldIf(generator, "postAggregations", postAggs.size() > 0 ? postAggs : null);
         writeField(generator, "intervals", intervals);
-        generator.writeNumberField("threshold", fetch);
+        generator.writeNumberField("threshold", fetch == null ? 0 : fetch);
```
too defensive. we would not enter this code if fetch is null. (probably also if collationIndexes is null.) so, better to replace with `assert fetch != null`

julianhyde commented on linq4j/src/main/java/org/apache/calcite/linq4j/LookupImpl.java:
```diff
@@ -88,8 +88,7 @@ public boolean containsKey(Object key) {
     List<V> list = (List<V>) value;
-    Enumerable<V> enumerable = Linq4j.asEnumerable(list);
-    return map.containsValue(enumerable);
+    return map.containsValue(list);
```
this method looks to be broken - it will always return false. `Linq4j.asEnumerable(List)` returns a new object that does not override `hashCode`. if we're going to fix it, let's add a test.

1m2c3t4 replied:
added test class LookupImplTest

julianhyde commented on core/src/main/java/org/apache/calcite/rel/rules/JoinToMultiJoinRule.java:
```diff
@@ -348,7 +348,7 @@ private void copyOuterJoinInfo(
       destJoinSpecs.addAll(srcJoinSpecs);
     } else {
-      int nFields = srcFields.size();
+      int nFields = srcFields == null ? 0 : srcFields.size();
```
a little defensive. `assert srcFields != null` would be better, since that is the implicit contract. also `assert destFields != null`

1m2c3t4 commented:
@julianhyde thanks for the review, and sorry for the delay in addressing the comments (I was ill, and have only recently recovered). In the mean time, it seems that DruidQuery has undergone substantial change, so the changes to that class no longer apply. I have addressed your other comments, removing the defensive coding and introducing asserts as suggested. I have also added a test class for LookupImpl.
